### PR TITLE
Add inverse scale and axpy

### PR DESCRIPTION
**Pull request by upsj**
This PR adds inv_scale and sub_scaled functions to Dense which will later be used to split the Arnoldi kernels in GMRES into individual dot products and orthogonalizations, as is necessary for a distributed implementation.

### CODE REVIEW

**Review by yhmtsai — APPROVED** [review]
LGTM. do you split gmres in this pr or another pr?

File: core/matrix/dense.cpp [hunk 1]
```diff
@@ -291,6 +308,29 @@ void Dense<ValueType>::add_scaled_impl(const LinOp *alpha, const LinOp *b)
...
+    auto exec = this->get_executor();
+
+    if (dynamic_cast<const Diagonal<ValueType> *>(b)) {
+        exec->run(dense::make_sub_scaled_diag(
+            as<Dense<ValueType>>(alpha),
+            dynamic_cast<const Diagonal<ValueType> *>(b), this));
+    } else {
+        exec->run(dense::make_sub_scaled(
+            make_temporary_conversion<ValueType>(alpha).get(),
+            make_temporary_conversion<ValueType>(b).get(), this));
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/matrix/dense.cpp:329]: Yes, make_temporary_conversion assumes the parameters are Dense of some value type

File: common/matrix/dense_kernels.cpp [hunk 2]
```diff
@@ -149,6 +199,24 @@ void add_scaled_diag(std::shared_ptr<const DefaultExecutor> exec,
 GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_DENSE_ADD_SCALED_DIAG_KERNEL);
 
 
+template <typename ValueType>
+void sub_scaled_diag(std::shared_ptr<const DefaultExecutor> exec,
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/matrix/dense_kernels.cpp:203]: I wanted to keep the symmetry, but we can also remove this if we consider it not useful

**Comment by upsj** [comment]
@yhmtsai Yes, I am trying to keep the PRs smaller. Refactoring GMRES should be another one.

**Review by hartwiganzt — APPROVED** [review]

**Comment by upsj** [comment]
rebase!

**Pull request merged**